debugger/Main.hs

```diff
@@ -21,19 +21,126 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+          , Option ['i'] []                   (ReqArg Datalog  "FILE")              "DDlog program to compile."
+          , Option ['L'] []                   (ReqArg LibDir   "PATH")              "Extra DDlog library directory."
+          , Option ['o'] ["output-dir"]       (ReqArg OutputDir "DIR")              "Output directory (default based on program name)."
+          , Option []    ["output-internal-relations"]  (NoArg OutputInternal)  "All non-input relations are marked as output relations."
+          , Option []    ["output-input-relations"]  (ReqArg OutputInput "PREFIX") "Mirror each input relation into an output relation named by prepending the prefix."
+          ]
+
+addOption :: Config -> TOption -> IO Config
+addOption config (DebugDumpFile f) = return config { confDebugDumpFile = f}
+addOption config (Datalog f)       = return config { confDatalogFile  = f}
+addOption config (LibDir d)        = return config { confLibDirs = nub (d:confLibDirs config)}
+addOption config (OutputDir d)     = return config { confOutputDir = d }
+addOption config Help              = return config { confAction = ActionHelp}
+addOption config OutputInternal   = return config { confOutputInternal = True }
+addOption config (OutputInput p)  = return config { confOutputInput = p }
+addOption config DumpFlat         = return config { confDumpFlat = True }
+
+validateConfig :: Config -> IO ()
+validateConfig Config{..} = do
+    when (confDatalogFile == "" && confAction /= ActionHelp && confAction /= ActionVersion)
+         $ errorWithoutStackTrace "input file not specified"
 
 main :: IO ()
 main = do
-       (filename:_) <- getArgs
-       contents <- readFile filename
-       let result = parseText contents
-       putStr result
+       args <- getArgs
+       prog <- getProgName
+       home <- lookupEnv "DDLOG_HOME"
+       config <- case getOpt Permute options args of
+                      (flags, [], []) -> do
+                                            conf <- foldM addOption defaultConfig flags
+                                            validateConfig conf
+                                            return conf
+                                         `catch`
+                                         (\e -> do putStrLn $ usageInfo ("Usage: " ++ prog ++ " [OPTION...]") options
+                                                   throw (e::SomeException))
+                      _ -> errorWithoutStackTrace $ usageInfo ("Usage: " ++ prog ++ " [OPTION...]") options
+       config' <- case home of
+           Just(p) -> addOption config (LibDir $ p ++ "/lib")
+           _       -> return config
+       do
+            datalogProg <- parseProgram config'
+            events <- parseEventFromDumpFile config'
+            let recordMap = handleDebugEvents events emptyDebuggerMaps datalogProg
+                s = queryAll events (dbgRecordMap recordMap)
+            putStr $ show s
+
+queryAll :: [Event] -> DebuggerRecordMap -> String
```

Comment:
So, currently the debugger tool will just print/query all derivations of all output records?

Not required for this change, but I wonder if we can make it more interactive (i.e., list out the output records, and then we can specify which output record we want to trace the derivation).

Reply:
Yeah, for now the debugger tool's function is 
queryDerivations :: DebuggerRecord -> DebuggerRecordMap -> DebuggerRecordNode
It takes one output record and debuggerRecordMap (got built when parsing log events) and return the root node of the derivations since the derivation result is a tree like structure.
According to Leonid, we are planning to design a UI that will take the debugger output and then display it in a better way, probably is a next step work.